### PR TITLE
Cgt opt

### PR DESCRIPTION
**Pull request by Sentient07**
Continuation of https://github.com/nouiz/Theano/pull/44

### CODE REVIEW

File: theano/gof/opt.py [hunk 1]
```diff
@@ -1283,8 +1288,8 @@ def apply_mult_opts(opt_list, node, multiple_opts=False):
                     new_node = repl[0].owner
                     apply_mult_opts(opt_list, new_node, True)
             return repl
-
-        return apply_mult_opts(self.opts, node, self.apply_all_opts)
+        opts = self.track_map.get(type(node.op), [])
```
Comment (nouiz) [inline theano/gof/opt.py]: You don't query it enough. YOu should add the next line:

opts += self.track_map.get(node.op, [])
opts += self.track_map.get(None, [])

Comment (Sentient07) [inline theano/gof/opt.py]: True, but our idea is to try lesser optimizations right ?

Comment (nouiz) [inline theano/gof/opt.py]: Yes, but currently it try not enough optimization. The goal is to skip optimization that won't get applied. Even with my version, this will happen.

Comment (Sentient07) [inline theano/gof/opt.py]: Why `None`. I didn't understand that ?

File: theano/tensor/basic.py [hunk 2]
```diff
@@ -2284,6 +2284,8 @@ def ones_like(model, dtype=None):
     """
     if dtype is None:
         dtype = model.type.dtype
+    if model.ndim == 0:
+        return constant(1.0, dtype=dtype)
```
Comment (Sentient07) [inline theano/tensor/basic.py]: This checks causes a problem. They return a constant when we use a scalar which results in `UnusedInputError` in many tests

Comment (Sentient07) [inline theano/tensor/basic.py]: In fact, all the errors Travis reports are because of returning constants except the ones that contain cycles.

File: theano/gof/opt.py (hunk 1 above)
Comment (nouiz) [inline theano/gof/opt.py]: When a local optimization want to run for all nodes, it register None in the tracks.

File: theano/tensor/basic.py (hunk 2 above)
Comment (nouiz) [inline theano/tensor/basic.py]: Ok. So remove that changes.

**Comment by Sentient07** [comment]
The cycle is introduced because of local_useless_elemwise optimizer, which happens to be at the Input of the graph.

**Comment by Sentient07** [comment]
The optimization done in l:2023 in `theano/tensor/opt.py` causes the cycle in graph. https://github.com/Sentient07/Theano/blob/addef9abab583dac8ccd1f759ea906ac99e415bd/theano/tensor/opt.py#L2023

I'm not able to understand more as to what `elemwise` operation it is trying to perform.

**Comment by nouiz** [comment]
I don't understand why this would cause problems. Can you replace that line by just:

return

and do you still have problems with it? If you don't add  local_useless_elemwise optimizer in useless, do you still have problems?

**Comment by Sentient07** [comment]
> I don't understand why this would cause problems. Can you replace that line by just:
> 
> return
This fixes it. I'm still checking if it causes problem elsewhere 
> and do you still have problems with it? If you don't add local_useless_elemwise optimizer in useless, do you still have problems?
This also fixes, but i'll resort to this if the above fix isn't compatible

File: theano/tensor/opt.py [hunk 3]
```diff
@@ -1901,6 +1877,7 @@ def local_subtensor_remove_broadcastable_index(node):
 
 @register_specialize
 @register_canonicalize('fast_compile_gpu')
+#@register_useless
```
Comment (Sentient07) [inline theano/tensor/opt.py]: You've purposefully left is commented or you wanted to add back later ?

File: theano/gof/opt.py [hunk 4]
```diff
@@ -1959,6 +1986,44 @@ def __str__(self):
                        '<TopoOptimizer instance>')
 
 
+def out2in(*local_opts, **kwargs):
```
Comment (Sentient07) [inline theano/gof/opt.py]: @nouiz Can i change all the imports of this module to this path ? You've added this in `theano/gof/opt.py` and that causes flake test errors

**Comment by Sentient07** [comment]
There is one error that i'm not able to see on Travis but that is happening locally. Could you please verify it, @nouiz 

```
======================================================================
FAIL: theano.tensor.tests.test_opt.test_local_merge_alloc
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/Ramana/projects/macvnev/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/Ramana/projects/theano/Theano/theano/tensor/tests/test_opt.py", line 6472, in test_local_merge_alloc
    assert len(topo) == 1
AssertionError: 
```

**Comment by nouiz** [comment]
There is a few failure in travis. You can fix this one yourself:

```
  File "/home/travis/build/nouiz/Theano/theano/tensor/tests/test_opt.py", line 32, in <module>
    from theano.tensor.opt import (
ImportError: cannot import name out2in
```

**Comment by Sentient07** [comment]
I had removed as flake complained. In the new commits i've changed the import path

File: theano/tensor/basic.py [hunk 5]
```diff
@@ -2315,6 +2315,8 @@ def zeros(shape, dtype=None):
         shape = [shape]
     if dtype is None:
         dtype = config.floatX
+    if isinstance(shape, (list, tuple)) and len(shape) == 0:
```
Comment (nouiz) [inline theano/tensor/basic.py]: finally, revert this as it cause a test failure:

```
======================================================================

ERROR: test_savemem_does_not_duplicate_number_of_scan_nodes (theano.scan_module.tests.test_scan.T_Scan)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/build/nouiz/Theano/theano/scan_module/tests/test_scan.py", line 4180, in test_savemem_does_not_duplicate_number_of_scan_nodes

    tmp_fn = theano.function([var], values)

  File "/home/travis/build/nouiz/Theano/theano/compile/function.py", line 322, in function

    output_keys=output_keys)

  File "/home/travis/build/nouiz/Theano/theano/compile/pfunc.py", line 391, in pfunc

    for p in params]

  File "/home/travis/build/nouiz/Theano/theano/compile/pfunc.py", line 485, in _pfunc_param_to_in

    raise TypeError('Constants not allowed in param list', param)

TypeError: ('Constants not allowed in param list', TensorConstant{1.0})

----------------------------------------------------------------------
```

idem bellow.

File: theano/tensor/opt.py [hunk 6]
```diff
@@ -2020,7 +2021,7 @@ def ones_like(node, in_idx):
             return [node.inputs[0]]
         elif (node.op.scalar_op == theano.scalar.identity and
               len(node.inputs) == 1):
-            return [node.inputs[0]]
+            return
```
Comment (nouiz) [inline theano/tensor/opt.py]: I don't like this fix. It disable an optimization. We need to find another fix.

Comment (Sentient07) [inline theano/tensor/opt.py]: Would cloning the variable and returning that be a good fix ?

**Comment by Sentient07** [comment]
> There is one error that i'm not able to see on Travis but that is happening locally. Could you please verify it, @nouiz

I now am able to see this error in Travis as well. I'll see which change brought this up

File: theano/tensor/opt.py (hunk 6 above)
Comment (Sentient07) [inline theano/tensor/opt.py]: I had tried removing this condition checking here. The tests passed now without causing cycles in the graph. Since the OP itself does what the optimization is trying to do, why can't we remove this condition (condition of `node.op.scalar_op == theano.scalar.identity`) ?

Comment (Sentient07) [inline theano/tensor/opt.py]: I've another question. When we're trying to replace an output variable with a variable that has no `var.owner`, is that the reason why the cycles are caused here ?

Comment (Sentient07) [inline theano/tensor/opt.py]: ^ That seems more logical to me but i get this error as well `RuntimeError: maximum recursion depth exceeded while calling a Python object`

Comment (nouiz) [inline theano/tensor/opt.py]: I think there is a bug deep in Theano that is triggered by this code in combination with the change in this PR.

I think it is related to your question, when we replace a variable by a variable that don't have an owner. I think the way we do the substitution is causing problems.

For the previous question, we absolutely need to keep that condition. Otherwise, we just remove the elemewise for other elemwise like tanh. So this case is only for the identity op.

Comment (Sentient07) [inline theano/tensor/opt.py]: > I think it is related to your question, when we replace a variable by a variable that don't have an owner. I think the way we do the substitution is causing problems.

How to do the replacement ? A simple `=` operator is creating problem of recursion depth

> For the previous question, we absolutely need to keep that condition. Otherwise, we just remove the elemewise for other elemwise like tanh. So this case is only for the identity op.

How is this condition in the optimization useful ? My question was, the Op and it's optimization both are doing the same right ? `Identity(Var)` would return the `Var` itself and here, the optimization is also trying to do the same. Or, what i'm thinking of how it works is wrong ?

**Comment by Sentient07** [comment]
By mistake, i added a few diff to your commit (	8a63457). That includes adding the decorator to the new optimizers that you've suggested in the email.

File: theano/tensor/opt.py (hunk 6 above)
Comment (nouiz) [inline theano/tensor/opt.py]: On Wed, Aug 3, 2016 at 1:26 PM, Ramana Subramanyam <notifications@github.com
> wrote:

> In theano/tensor/opt.py
> <https://github.com/nouiz/Theano/pull/45#discussion_r73372841>:
>
> > @@ -2020,7 +2021,7 @@ def ones_like(node, in_idx):
> >              return [node.inputs[0]]
> >          elif (node.op.scalar_op == theano.scalar.identity and
> >                len(node.inputs) == 1):
> > -            return [node.inputs[0]]
> > +            return
>
> I think it is related to your question, when we replace a variable by a
> variable that don't have an owner. I think the way we do the substitution
> is causing problems.
>
> How to do the replacement ? A simple = operator is creating problem of
> recursion depth
>

Keep this diff in the PR. I don't have time to explain all the replacement
code. It isn't well documented, so if you can't understand that code, I'll
just fix this one myself.

Keep this diff in the PR and make sure there isn't other error in Theano
tests due to that PR.


> For the previous question, we absolutely need to keep that condition.
> Otherwise, we just remove the elemewise for other elemwise like tanh. So
> this case is only for the identity op.
>
> How is this condition in the optimization useful ? My question was, the Op
> and it's optimization both are doing the same right ? Identity(Var) would
> return the Var itself and here, the optimization is also trying to do the
> same. Or, what i'm thinking of how it works is wrong ?
>

Here the optimization remove Identify from the graph. So this speed up
computaion.

File: theano/compile/mode.py [hunk 7]
```diff
@@ -181,6 +181,18 @@ def apply(self, fgraph):
 optdb.register('merge1', gof.MergeOptimizer(),
                0, 'fast_run', 'fast_compile', 'merge')
 
+local_useless = gof.optdb.LocalGroupDB(apply_all_opts=True)
+optdb.register('useless', gof.optdb.TopoDB(local_useless),
+               0.6, 'fast_run', 'fast_compile')
+
+# After scan1 opt at 0.5 and before ShapeOpt at 1
```
Comment (nouiz) [inline theano/compile/mode.py]: Move all this comment above the local_useless opt

File: theano/compile/profiling.py [hunk 8]
```diff
@@ -54,7 +54,7 @@ def _atexit_print_fn():
         destination_file = open(config.profiling.destination, 'w')
 
     for ps in _atexit_print_list:
-        if ps.fct_callcount >= 1 or ps.compile_time > 1:
+        if ps.fct_callcount > 1 or ps.compile_time > 1:
```
Comment (nouiz) [inline theano/compile/profiling.py]: remove that from this PR.

Comment (Sentient07) [inline theano/compile/profiling.py]: This file isn't on my tracked list of git files. I'm not sure how it is occuring

File: theano/gof/opt.py [hunk 9]
```diff
@@ -1237,16 +1237,25 @@ class LocalOptGroup(LocalOptimizer):
...
             optimizers = tuple(optimizers[0])
         self.opts = optimizers
+        assert isinstance(self.opts, tuple)
+
         self.reentrant = any(getattr(opt, 'reentrant', True)
                              for opt in optimizers)
         self.retains_inputs = all(getattr(opt, 'retains_inputs', False)
                                   for opt in optimizers)
 
+        self.apply_all_opts = kwargs.get('apply_all_opts', False)
```
Comment (nouiz) [inline theano/gof/opt.py]: replace get by pop.

then add this assert len(kwargs) == 0. So if people make typo or add other parameter, it will raise an error.

File: theano/gof/opt.py [hunk 10]
```diff
@@ -1237,16 +1237,25 @@ class LocalOptGroup(LocalOptimizer):
```
Comment (nouiz) [inline theano/gof/opt.py]: It would be good if you fill this at the same time.

File: theano/gof/opt.py [hunk 11]
```diff
@@ -1261,10 +1270,28 @@ def tracks(self):
...
+                if not repl:
+                    continue
+                else:
+                    if not multiple_opts or not repl[0].owner:
+                        return repl
+                    assert len(repl) == 1
+                    # Ensuring not the input of graph
+                    assert repl[0].owner
+                    new_node = repl[0].owner
+                    apply_mult_opts(opt_list, new_node, True)
```
Comment (nouiz) [inline theano/gof/opt.py]: Here, you must compute the new opt_list, as the node can be different.

File: theano/gof/opt.py [hunk 12]
```diff
@@ -1959,6 +1986,44 @@ def __str__(self):
                        '<TopoOptimizer instance>')
 
 
+def out2in(*local_opts, **kwargs):
+    """WRITEME """
```
Comment (nouiz) [inline theano/gof/opt.py]: Fill the WRITEME at the same time.

File: theano/gof/opt.py [hunk 13]
```diff
@@ -1959,6 +1986,44 @@ def __str__(self):
...
+                        order='out_to_in',
+                        failure_callback=TopoOptimizer.warn_inplace,
+                        **kwargs)
+    if name:
+        ret.__name__ = name
+    return ret
+
+
+def in2out(*local_opts, **kwargs):
+    """WRITEME """
```
Comment (nouiz) [inline theano/gof/opt.py]: idem.

File: theano/gof/optdb.py [hunk 14]
```diff
@@ -399,11 +402,41 @@ class LocalGroupDB(SequenceDB):
...
+        # For the new `useless` optimizer
+        opts = super(LocalGroupDB, self).query(*tags, **kwtags)
+        ret = opt.LocalOptGroup(*opts, apply_all_opts=self.apply_all_opts)
+        return ret
+
+
+class TopoDB(DB):
+    """
+
+    Generate a local optimizer of type LocalOptGroup instead
```
Comment (nouiz) [inline theano/gof/optdb.py]: Update this. It generate an TopoOptimizer, not a LocalOptGroup

File: theano/tensor/opt.py [hunk 15]
```diff
@@ -1755,9 +1731,7 @@ def local_useless_fill(node):
             return [v]
 
 
-@register_specialize
-@register_stabilize
-@register_canonicalize
```
Comment (nouiz) [inline theano/tensor/opt.py]: keep the previous register.

Comment (Sentient07) [inline theano/tensor/opt.py]: Oh!! Which commit removed it ? Was it your yesterday's ? I don't remember doing this change...

File: theano/tensor/opt.py [hunk 16]
```diff
@@ -1901,6 +1904,7 @@ def local_subtensor_remove_broadcastable_index(node):
 
 @register_specialize
 @register_canonicalize('fast_compile_gpu')
+# @register_useless
```
Comment (nouiz) [inline theano/tensor/opt.py]: Do you remember why this one was commented?

if not, uncomment it and see if it cause more failure. If not, uncomment. Otherwise, show me the errors.

File: theano/tensor/opt.py [hunk 17]
```diff
@@ -4683,6 +4707,10 @@ def transform(self, node):
         assert len(node.outputs) == 1
         out = node.outputs[0]
 
+        # Condition for replacement variable not being a part of the graph
+        if not hasattr(out, 'clients'):
+            return False
+
```
Comment (nouiz) [inline theano/tensor/opt.py]: remove this. I think it is a left over from another fix. If removing this cause errors, tell me which one.

File: theano/tensor/opt.py (hunk 15 above)
Comment (nouiz) [inline theano/tensor/opt.py]: I don't know, just revert it.

File: theano/tensor/opt.py (hunk 16 above)
Comment (Sentient07) [inline theano/tensor/opt.py]: This was commented from the first itself. I asked you about this previously, and you had asked me to keep it as it is. I removed it and it didn't cause any errors as well

File: theano/gof/optdb.py [hunk 18]
```diff
@@ -417,10 +417,7 @@ def query(self, *tags, **kwtags):
 class TopoDB(DB):
     """
 
-    Generate a local optimizer of type LocalOptGroup instead
-    of a global optimizer.
-
-    It supports the tracks, to only get applied to some Op.
+    Generate a local optimizer of type TopoOptimizer.
```
Comment (nouiz) [inline theano/gof/optdb.py]: It generate a global optimization.

File: theano/gof/opt.py [hunk 19]
```diff
@@ -1233,8 +1238,16 @@ def decorator(f):
...
     """
-    WRITEME
+    Takes a list of LocalOptimizer and applies them to the nodes.
+    When apply_all_opts is set to True, it tries multiple optimization to a node.
+    A node is first optimized and the list of optimizers for the optimized node is applied
 
+    Parameters
+    ----------
+    optimizers : List
+    List of optimizers to be applied to a node
```
Comment (nouiz) [inline theano/gof/opt.py]: I think this line must be indented by 4 spaces.

File: theano/gof/opt.py [hunk 20]
```diff
@@ -201,7 +201,12 @@ def warn(exc, self, optimizer):
 
     def __init__(self, *opts, **kw):
         """
-        WRITEME
+        Parameters
+        ----------
+        opts : List
+        The List of optimizers to be applied to a node
```
Comment (nouiz) [inline theano/gof/opt.py]: I think the indentation is wrong.

File: theano/gof/opt.py [hunk 21]
```diff
@@ -201,7 +201,12 @@ def warn(exc, self, optimizer):
...
+        kw : Dict
+        Dictonary containing failure call back message
```
Comment (nouiz) [inline theano/gof/opt.py]: Tell that the only supported keyword is failure_callback. It is not a message but a callback when an error happen.

File: theano/gof/opt.py [hunk 22]
```diff
@@ -1233,8 +1238,16 @@ def decorator(f):
 
 class LocalOptGroup(LocalOptimizer):
     """
-    WRITEME
+    Takes a list of LocalOptimizer and applies them to the nodes.
+    When apply_all_opts is set to True, it tries multiple optimization to a node.
+    A node is first optimized and the list of optimizers for the optimized node is applied
```
Comment (nouiz) [inline theano/gof/opt.py]: What about this paragraph:

Apply many local optimizer on a node. If apply_all_opts is False, it will return after the first optimizer applied. Otherwise, it will start again with the node returned by an optimizer.

**Comment by nouiz** [comment]
/home/travis/build/nouiz/Theano/theano/gof/opt.py:2010:4: W291 trailing whitespace

File: theano/compile/mode.py [hunk 23]
```diff
@@ -187,8 +187,11 @@ def apply(self, fgraph):
 # New nodes that don't have infer_shape need that the original node
 # also don't have infer_shape
 local_useless = gof.optdb.LocalGroupDB(apply_all_opts=True)
-optdb.register('useless', gof.optdb.TopoDB(local_useless),
-               0.6, 'fast_run', 'fast_compile')
+optdb.register(
+    'useless',
+    gof.optdb.TopoDB(local_useless,
+                     failure_callback=gof.opt.NavigatorOptimizer.warn_inplace),
```
Comment (Sentient07) [inline theano/compile/mode.py]: It's a little surprising :) But how did this fix the error ?

Comment (nouiz) [inline theano/compile/mode.py]: If in the user graph, there is inplace node, then we need to use the inplace failure call back. User don't do that. We don't support it well when user do that. But some tests need to do that to test well.

File: theano/gof/opt.py [hunk 24]
```diff
@@ -1309,17 +1322,48 @@ def apply_mult_opts(opt_list, node, multiple_opts=False):
...
         if depth != 0:
             depth -= 1
             for lopt in self.opts:
                 lopt.print_summary(stream, level=(level + 2), depth=depth)
 
+
+        count_opt = []
+        not_used = []
+        for o, count in iteritems(self.process_count):
+            if count > 0:
```
Comment (Sentient07) [inline theano/gof/opt.py]: Neither the if or the else statement gets executed. The function exits from here. This was what i was able to see from pdb

Comment (Sentient07) [inline theano/gof/opt.py]: I realize that dictionary is empty because of an error i made.

File: theano/gof/opt.py [hunk 25]
```diff
@@ -1296,10 +1300,19 @@ def compute_opts(node):
...
                 repl = opt.transform(node)
+                opt_finish = time.time()
                 if not repl:
                     continue
                 else:
+                    self.time_opts[opt] = opt_start - opt_finish
+                    try:
+                        self.node_created[opt] += len(graph.ops(node.inputs, node.outputs))
+                        self.process_count[opt] += 1
+                    except KeyError:
```
Comment (Sentient07) [inline theano/gof/opt.py]: This is the part that caused error in this commit. I've addressed in the upcoming commit.

File: theano/gof/opt.py [hunk 26]
```diff
@@ -1309,17 +1322,48 @@ def apply_mult_opts(opt_list, node, multiple_opts=False):
                     new_opts = compute_opts(new_node)
                     apply_mult_opts(new_opts, new_node, True)
             return repl
-
-        return apply_mult_opts(compute_opts(node), node, self.apply_all_opts)
+        node_start = time.time()
```
Comment (Sentient07) [inline theano/gof/opt.py]: In this i'm computing only the nodes optimized initially

File: theano/gof/opt.py [hunk 27]
```diff
@@ -1261,19 +1288,82 @@ def tracks(self):
...
 
     def transform(self, node):
-        for opt in self.opts:
-            repl = opt.transform(node)
-            if repl:
-                return repl
+        if len(self.opts) == 0:
+            return
+
+        def compute_opts(node):
```
Comment (nouiz) [inline theano/gof/opt.py]: instead of having a function compute_opts, can you move that code inside apply_mult_opts, just before the loop? This will save a function call and would make the code easier to understand.

File: theano/gof/optdb.py [hunk 28]
```diff
@@ -321,8 +321,11 @@ def __init__(self, failure_callback=opt.SeqOptimizer.warn):
 
     def register(self, name, obj, position, *tags):
         super(SequenceDB, self).register(name, obj, *tags)
-        assert isinstance(position, (integer_types, float))
-        self.__position__[name] = position
+        if position == 'last':
+            self.__position__[name] = len(self.__position__)
```
Comment (nouiz) [inline theano/gof/optdb.py]: this should be:

self.__position__[name] = max(self.__position__.values())

check, maybe for python 3 you need to get the values of the dict differently.

File: theano/tensor/opt.py [hunk 29]
```diff
@@ -1758,6 +1734,7 @@ def local_useless_fill(node):
 @register_specialize
 @register_stabilize
 @register_canonicalize
+@register_useless
```
Comment (nouiz) [inline theano/tensor/opt.py]: Here include just in useless, so remove register_specialize, stabilize and canonicalize.

File: theano/tensor/opt.py [hunk 30]
```diff
@@ -2528,6 +2538,7 @@ def local_useless_slice(node):
             return [out]
 
 
+@register_useless
```
Comment (nouiz) [inline theano/tensor/opt.py]: remove. useless is executed before the shape_feature is added. So the condition at line 2559 will always fail.

File: theano/tensor/opt.py [hunk 31]
```diff
@@ -2969,6 +2985,7 @@ def local_subtensor_merge(node):
...
 @register_canonicalize
 @register_specialize
 @gof.local_optimizer([Subtensor])
```
Comment (nouiz) [inline theano/tensor/opt.py]: Can you change line 2976 to those lines:
"""
alloc(val)[x:y] -> alloc(val[...])
alloc(val)[x:y] -> alloc(val)
This can be seen as a lift, but it also reduce the number of computation/memory.
"""

I would keep it in the useless. Maybe this is the optimization that make the graph bigger. This need to be validated.

File: theano/gof/opt.py (hunk 24 above)
Comment (Sentient07) [inline theano/gof/opt.py]: The dict isn't empty but it has the count as the default one, 0. Could you please check this @nouiz ?

File: theano/gof/opt.py [hunk 32]
```diff
@@ -1261,19 +1291,80 @@ def tracks(self):
...
+
+            for opt in opt_list:
+                opt_start = time.time()
+                repl = opt.transform(node)
+                opt_finish = time.time()
+                if not repl:
+                    continue
+                else:
+                    self.time_opts[opt] = opt_start - opt_finish
+                    self.node_created[opt] += len(graph.ops(node.inputs, node.outputs))
```
Comment (nouiz) [inline theano/gof/opt.py]: ops(node.inputs, repl)

Comment (nouiz) [inline theano/gof/opt.py]: in fact, it should be:

ops(fgraph.variable, node.outputs)

where fgraph is node.fgraph from the original node. Maybe you need to pass fgraph to apply_mult_opts

**Comment by nouiz** [comment]
travis is failing with this error:

/home/travis/build/nouiz/Theano/theano/gof/opt.py:2010:4: W291 trailing whitespace

**Comment by Sentient07** [comment]
Yeah, i still have to address your yesterday's comment and travis.

File: theano/gof/opt.py [hunk 33]
```diff
@@ -1328,27 +1336,27 @@ def apply_mult_opts(opt_list, node, multiple_opts=False):
...
-            depth -= 1
-            for lopt in self.opts:
-                lopt.print_summary(stream, level=(level + 2), depth=depth)
-
-
+    @staticmethod
+    def print_profile(stream, prof, level=0):
+        (time_opts, time_nodes, node_created, process_count) = prof
+        import pdb
+        pdb.set_trace()
```
Comment (Sentient07) [inline theano/gof/opt.py]: I never got to this place.
@lamblin

File: theano/gof/opt.py [hunk 34]
```diff
@@ -1297,6 +1290,8 @@ def transform(self, node):
     # This method is just kept for the `print_profile` method to know the variables that
     # needs to be printed
     def apply(self):
+        import pdb
+        pdb.set_trace()
```
Comment (Sentient07) [inline theano/gof/opt.py]: Neither i got to debug from here
@lamblin

File: theano/gof/opt.py [hunk 35]
```diff
@@ -2038,16 +2039,19 @@ def pruner(node):
...
         (opt, nb, nb_nodes_start, nb_nodes_end,
-         io_t, loop_t, callback_time) = prof
+         io_t, loop_t, callback_time, lopt) = prof
 
         blanc = ('    ' * level)
         print(blanc, "TopoOptimizer ",
               getattr(opt, "name", getattr(opt, "__name__", "")), file=stream)
+        if getattr(opt, "name", getattr(opt, "__name__", "")) == 'useless':
+            import pdb
+            pdb.set_trace()
```
Comment (Sentient07) [inline theano/gof/opt.py]: when i called the methods apply and passed it's parameters to  `print_profile` from here, i just got to see "LocalOptGroup" and when i checked with time_opts dictionary, i found it to be having all the values as 0.

File: theano/gof/opt.py (hunk 32 above)
Comment (Sentient07) [inline theano/gof/opt.py]: Yeah! It's complicated to do that from LocalOptimizer. so @lamblin had asked me to remove printing the node_created for this.

File: theano/gof/opt.py (hunk 27 above)
Comment (Sentient07) [inline theano/gof/opt.py]: I just noticed that doing the computation the way you told is costing extra 40sec in profiling. I'm not able to understand why that happens. But when i checked, it was this commit that caused the error.

**Comment by Sentient07** [comment]
There is this message in travis at the end of second build.
```checks the gradient for the case ... /home/travis/build.sh: line 45: 10092 Killed theano-nose -v $PART```
Also, the second link shows failure, but i'm not able to see why it is failing.

File: theano/gof/opt.py [hunk 36]
```diff
@@ -1291,35 +1291,30 @@ def transform(self, node):
...
+            opts = []
+            opts.append(self.track_map.get(type(node.op), []) + self.track_map.get(node.op, []) + self.track_map.get(None, []))
 
-            for opt in opt_list:
+            for opt in opts:
                 opt_start = time.time()
                 repl = opt.transform(node)
                 opt_finish = time.time()
+                self.time_opts[opt] = opt_start - opt_finish
+                self.process_count[opt] += 1
```
Comment (nouiz) [inline theano/gof/opt.py]: This one should be in the else clause. It is to count the number of time an opt applied. Not the number of time we try an optimization.

Comment (Sentient07) [inline theano/gof/opt.py]: Yes, but you had asked me to move it from the else before.

Comment (Sentient07) [inline theano/gof/opt.py]: Or, do i need to maintain another variable for printing ? 
https://github.com/nouiz/Theano/pull/45#commitcomment-18548447

File: theano/gof/optdb.py [hunk 37]
```diff
@@ -322,7 +322,7 @@ def __init__(self, failure_callback=opt.SeqOptimizer.warn):
     def register(self, name, obj, position, *tags):
         super(SequenceDB, self).register(name, obj, *tags)
         if position == 'last':
-            self.__position__[name] = len(self.__position__)
+            self.position[name] = max(self.position.values())
```
Comment (nouiz) [inline theano/gof/optdb.py]: Why did you change `__position__` to `position`? Probably due to a github highlight.

Comment (Sentient07) [inline theano/gof/optdb.py]: You had asked me to make this change :) 
https://github.com/nouiz/Theano/pull/45/files/08a05156eaf3b32d43426a689b3d3228fe01771b#r73763761

File: theano/gof/opt.py (hunk 32 above)
Comment (nouiz) [inline theano/gof/opt.py]: Did you found why there is more nodes after the useless pass?

File: theano/gof/opt.py (hunk 27 above)
Comment (nouiz) [inline theano/gof/opt.py]: I'll comment in the new code.

File: theano/gof/optdb.py (hunk 37 above)
Comment (nouiz) [inline theano/gof/optdb.py]: It was github syntax highlight that caused this Here I wrote the the good with without back-quote: __position__ it get displayed as bold by github. If I back-quote it, then it get well displayed: `__position__`

File: theano/gof/opt.py [hunk 38]
```diff
@@ -1261,10 +1293,70 @@ def tracks(self):
...
-        for opt in self.opts:
-            repl = opt.transform(node)
-            if repl:
-                return repl
+        if len(self.opts) == 0:
+            return
+
+        def apply_mult_opts(node, multiple_opts=False):
+            repl = False
+            opts = self.track_map.get(type(node.op), []) + self.track_map.get(node.op, []) + self.track_map.get(None, [])
```
Comment (nouiz) [inline theano/gof/opt.py]: I did some timing. It would be faster to use a default dict. Use this code instead (place at the right lines):

from collections import defaultdict

self.track_map=collections.defaultdict(lambda: [])

self.track_map[c].append(o)

opts = self.track_map[type(node.op)] + self.track_map[node.op] + self.track_map[None]

Can you time again with your use case? I timed the dict operation only.

**Comment by nouiz** [comment]
It is probably the break point that caused problems in travis.

File: theano/gof/opt.py (hunk 36 above)
Comment (nouiz) [inline theano/gof/opt.py]: I think @lamblin idea to have both is a good idea.

File: theano/gof/opt.py [hunk 39]
```diff
@@ -1309,24 +1313,27 @@ def apply_mult_opts(node, multiple_opts=False):
...
                     if not multiple_opts or not repl[0].owner:
                         return repl
                     assert len(repl) == 1
                     # Ensuring not the input of graph
                     assert repl[0].owner
                     new_node = repl[0].owner
-                    apply_mult_opts(new_node, True)
+                    if hasattr(new_node, 'fgraph'):
+                        apply_mult_opts(new_node, new_node.fgraph, True)
+                    apply_mult_opts(new_node, False, True)
```
Comment (nouiz) [inline theano/gof/opt.py]: this is bad, if new_node have fgraph, then apply_mult_opts will be called twice on the new node. Also, you should return the result of apply_mult_opts. I think those 3 lines should be:

So we will keep the fgraph and be able to compute "something" even when new_nodes don't have an fgraph.

                        return apply_mult_opts(new_node, fgraph, True)

@lamblin, I would prefer to do it like this then not doing anything. The "something" perfect. It don't show the number of nodes removed. Some opt don't create new nodes, just remove some nodes. Also, the second and following opt will get the number of nodes as the total for the optimization and the previous one. Why I want to keep it? If multiple_opts is False, then the number of new nodes will be fine. It will also help guide which optimization create new nodes.

But to warn about that, I would add in the print_profile() method a note that this is an approximation of nodes created when multiple_opts is True.

File: theano/tensor/opt.py [hunk 40]
```diff
@@ -1737,7 +1712,7 @@ def local_fill_to_alloc(node):
                        1.52, 'fast_run')
 
 
-@register_canonicalize("fast_compile")
+@register_useless
```
Comment (nouiz) [inline theano/tensor/opt.py]: do not remove the register_canonicalize that was there.

File: theano/gof/opt.py (hunk 36 above)
Comment (Sentient07) [inline theano/gof/opt.py]: Just to confirm, implemented this in the next commit.

**Comment by Sentient07** [comment]
This commit , 123878e2a288bf18718658a7687018df4476fd45 made me some ~5sec speed up. I would like to implement it elsewhere as well, mainly in Equilibrium Opt and GraphToGpu Opt. I'll open a new issue for this.

File: theano/gof/opt.py [hunk 41]
```diff
@@ -1313,7 +1304,7 @@ def apply_mult_opts(node, fgraph, multiple_opts=False):
                 if not repl:
                     continue
                 else:
-                    self.node_created[opt] += len(graph.ops(fgraph.variables, node.outputs))
+                    self.node_created[opt] += len(graph.ops(node.inputs, repl))
```
Comment (nouiz) [inline theano/gof/opt.py]: I think it should be len(graph.ops(fgraph.variables, repl)).

I think you find the problem we should not use node.output, but repl. But we should not use node.inputs, but fgraph.variables.

File: theano/gof/opt.py [hunk 42]
```diff
@@ -1304,16 +1304,14 @@ def apply_mult_opts(node, fgraph, multiple_opts=False):
...
                     self.applied_true[opt] += 1
                     if not multiple_opts or not repl[0].owner:
                         return repl
                     assert len(repl) == 1
                     # Ensuring not the input of graph
                     assert repl[0].owner
                     new_node = repl[0].owner
-                    if hasattr(new_node, 'fgraph'):
-                        apply_mult_opts(new_node, new_node.fgraph, True)
                     apply_mult_opts(new_node, fgraph, True)
```
Comment (nouiz) [inline theano/gof/opt.py]: I think you need to fix something here. I think it should be "return apply..."

File: theano/gof/opt.py (hunk 41 above)
Comment (Sentient07) [inline theano/gof/opt.py]: Yeah, i reverted it back in a commit after this

File: theano/gof/opt.py [hunk 43]
```diff
@@ -1312,8 +1312,9 @@ def apply_mult_opts(node, fgraph, multiple_opts=False):
                     # Ensuring not the input of graph
                     assert repl[0].owner
                     new_node = repl[0].owner
-                    apply_mult_opts(new_node, fgraph, True)
+                    apply_mult_opts(new_node, fgraph, True, repl)
```
Comment (nouiz) [inline theano/gof/opt.py]: Your fix isn't good. If the second opt get applied, it still won't be returned.

Comment (Sentient07) [inline theano/gof/opt.py]: If the second opt gets applied, there are two conditions, one is it returns the new outputs or it goes for optimizing the optimized node.(the third time). If nothing works in the third time, before, the code returned None/False , now it returns the previously optimized node's outputs .

Comment (nouiz) [inline theano/gof/opt.py]: Now it return the previously optized node, but it don't return the second
optimized node if 2 opt get applied.

**Comment by nouiz** [comment]
continued in https://github.com/Theano/Theano/pull/4876
